redskyctl/internal/commands/initialize/filters.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2020 GramLabs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initialize
+
+import (
+	"bytes"
+
+	"github.com/redskyops/redskyops-controller/internal/version"
+	"sigs.k8s.io/kustomize/api/filters/labels"
+	"sigs.k8s.io/kustomize/api/resid"
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// labelFilter returns a filter that applies the configured labels
+func (o *GeneratorOptions) labelFilter() kio.Filter {
+	f := labels.Filter{
+		Labels: map[string]string{
+			"app.kubernetes.io/version": version.GetInfo().Version,
+		},
+		FsSlice: []types.FieldSpec{
+			{
+				Gvk: resid.Gvk{
+					Kind: "Deployment",
+				},
+				Path:               "spec/template/metadata/labels",
+				CreateIfNotPresent: true,
+			},
+			{
+				Path:               "metadata/labels",
+				CreateIfNotPresent: true,
+			},
+		},
+	}
+
+	for k, v := range o.labels {
+		f.Labels[k] = v
+	}
+
+	return f
+}
+
+// clusterRoleBindingFilter returns a filter that removes cluster role bindings for namespaced deployments
+func (o *GeneratorOptions) clusterRoleBindingFilter() kio.Filter {
+	return kio.FilterFunc(func(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+		output := make([]*yaml.RNode, 0, len(nodes))
+		for _, n := range nodes {
+			m, err := n.GetMeta()
+			if err != nil {
+				return nil, err
+			}
+
+			if m.Kind == "ClusterRoleBinding" && m.APIVersion == "rbac.authorization.k8s.io/v1" {
+				continue
+			}
+
+			output = append(output, n)
+		}
+		return output, nil
+	})
+}
+
+// envFromSecretFilter comments out the `envFrom` field of the deployment
+func (o *GeneratorOptions) envFromSecretFilter() kio.Filter {
+	return kio.FilterFunc(func(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+		for _, n := range nodes {
+			m, err := n.GetMeta()
+			if err != nil {
+				return nil, err
+			}
+
+			if m.Kind != "Deployment" || m.Name != "redsky-controller-manager" {
+				continue
+			}
+
+			if err := commentOutEnvFrom(n); err != nil {
+				return nil, err
+			}
+		}
+		return nodes, nil
+	})
+}
+
+func commentOutEnvFrom(node *yaml.RNode) error {
```

Comment:
This was a pain, but I felt it produced a much more desirable outcome: instead of completely removing the reference to the secret, it is commented out. This is primarily for doing `redskyctl generate install --skip-secret` followed by `redskyctl generate secret` (once you generate the secret, you need to manually uncomment the reference to it). The alternative was to add an `optional: true` when using `--skip-secret` (like we do with the Helm chart), however that seem to be less reliable (and technically, `--skip-secret` _is_ undocumented).

Reply:
Fair enough, my concern is that actually including the `envFrom` will then conflict with the patch generated by `redskyctl authorize-cluster` to add the secret reference to the deployment later.

I tried to clean up the commenting code a bit, but maybe I also need to fix the patch on `authorize-cluster` so we can do an empty secret.

Reply:
Actually, it looks like there is no merge key on `envFrom` so the `authorize-cluster` smp does the right thing (relatively speaking, there is an unrelated bug with metadata though: PR incoming on that front)